// use-countdown-timer/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var react = require('react');

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function reducer(state, action) {
    switch (action.type) {
        case 'START':
            return __assign(__assign({}, state), { canStart: state.countdown !== 0 });
        case 'RESET':
            return __assign(__assign({}, state), { countdown: action.payload, canStart: false, isRunning: false });
        case 'PAUSE':
            return __assign(__assign({}, state), { canStart: false, isRunning: false });
        case 'RUNNING':
            return __assign(__assign({}, state), { isRunning: true });
        case 'TICK':
            return __assign(__assign({}, state), { countdown: state.countdown - action.payload });
        /* istanbul ignore next */
        default:
            return state;
    }
}
/**
 * Create a configurable countdown timer.
 */
function useCountdownTimer(_a) {
    var timer = _a.timer, _b = _a.interval, interval = _b === void 0 ? 1000 : _b, _c = _a.autostart, autostart = _c === void 0 ? false : _c, _d = _a.expireImmediate, expireImmediate = _d === void 0 ? false : _d, _e = _a.resetOnExpire, resetOnExpire = _e === void 0 ? true : _e, onExpire = _a.onExpire, onReset = _a.onReset;
    var _f = react.useReducer(reducer, {
        canStart: autostart,
        countdown: timer,
        isRunning: false,
    }), state = _f[0], dispatch = _f[1];
    function start() {
        dispatch({ type: 'START' });
    }
    function pause() {
        dispatch({ type: 'PAUSE' });
    }
    function initStopped(time) {
        dispatch({ type: 'RESET', payload: time });
    }
    var reset = react.useCallback(function () {
        initStopped(timer);
        if (onReset && typeof onReset === 'function') {
            onReset();
        }
    }, [timer, onReset]);
    var expire = react.useCallback(function () {
        initStopped(resetOnExpire ? timer : 0);
        if (onExpire && typeof onExpire === 'function') {
            onExpire();
        }
    }, [timer, onExpire, resetOnExpire]);
    react.useEffect(function () {
        function tick() {
            if (state.countdown / 1000 <= 0 ||
                (expireImmediate && (state.countdown - interval) / 1000 <= 0)) {
                expire();
            }
            else {
                dispatch({ type: 'TICK', payload: interval });
            }
        }
        var id;
        if (state.canStart) {
            id = setInterval(tick, interval);
            if (!state.isRunning) {
                dispatch({ type: 'RUNNING' });
            }
        }
        return function () { return clearInterval(id); };
    }, [
        expire,
        expireImmediate,
        interval,
        state.canStart,
        state.countdown,
        state.isRunning,
    ]);
    return {
        countdown: state.countdown,
        isRunning: state.isRunning,
        start: start,
        reset: reset,
        pause: pause,
    };
}

exports.useCountdownTimer = useCountdownTimer;
